game.py

```diff
@@ -1,18 +1,26 @@
-from random import randint
+from random import choice, randint
 
 from exceptions import NimException
 
 
 class Bot():
     def move(self, state):
-        raise NotImplementedError
+        valid_pile_indexes = [i for i, pile in enumerate(state) if pile >= 1]
```

Comment:
This comprehension appears multiple times the exact same way. This should always be a red flag. Look up what it means for code to be "D.R.Y." (don't repeat yourself) We should add a `valid_pile_indexes` method to the bot class that just returns `[i for i, pile in enumerate(state) if pile >= 1]`. Then, you can call it here as `if not self.valid_pile_indexes(state):` and, since it will be inherited by all subclasses, you can call it in you random bot as `pile_index = choice(self.get_valid_pile_indexes(state))`.

Reply:
I like learning these efficiencies, everything just cleans up over time.